File: LeetCodeOnlineJudge/3Sum/Solution.java
//Given an array S of n integers, are there elements a, b, c in S such that a + b + c = 0? Find all unique triplets in the array which gives the sum of zero.
import java.util.*;
public class Solution{
    public ArrayList<ArrayList<Integer>> threeSum(int[] num) {
        ArrayList<ArrayList<Integer>> ret = new ArrayList<ArrayList<Integer>>();
        if(num.length < 3) return ret;
        int len = num.length;
        int i = 0;
        int j = 0;
        int k = 0;
        Arrays.sort(num);
        for(i = 0; i < len - 2; i++){
            if(i > 0 && num[i] == num[i-1]) continue;
            j = i + 1;
            k = len - 1;
            while(j < k){
                if(num[k] == -1 * (num[i] + num[j])){
                    ArrayList<Integer> e = new ArrayList<Integer>();
                    e.add(num[i]);
                    e.add(num[j]);
                    e.add(num[k]);
                    ret.add(e);
                    k--;
                    j++;
                    while(k > j && num[k] == num[k+1]) k--;
                    while(j < k && num[j] == num[j-1]) j++;
                }else if(num[k] > -1 * (num[i] + num[j])){
                    k--;
                }else{
                    j++;
                }
            }
        }
        return ret;
    }
    public static void main(String args[]){
        Solution sol = new Solution();
        int A[] = {0,0,0};  
        ArrayList<ArrayList<Integer>> ret = sol.threeSum(A);
        for(ArrayList<Integer> e : ret){
            System.out.println(e);
        }
    }
}

File: LeetCodeOnlineJudge/3Sum/Solution.cpp
//Given an array S of n integers, are there elements a, b, c in S such that a + b + c = 0? Find all unique triplets in the array which gives the sum of zero.
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;
class Solution{
    public:
        vector<vector<int> > threeSum(vector<int> &num) {
            sort(num.begin(), num.end());
            int n = num.size();
            vector<vector<int> > ret;
            vector<int> curr(3);
            for(int i = 0; i < n - 2; i++){
                if(i > 0 && num[i] == num[i-1]) continue;
                int j = i + 1;
                int k = n - 1;
                int goal = 0 - num[i];
                curr[0] = num[i];
                while(j < k){
                    while(j < k && num[j] + num[k] < goal) j++;
                    while(j < k && num[j] + num[k] > goal) k--;
                    if(j < k && num[j] + num[k] == goal){
                        curr[1] = num[j];
                        curr[2] = num[k];
                        ret.push_back(curr);
                        j++;
                        k--;
                        while(j < k && num[j] == num[j-1]) j++;
                        while(j < k && num[k] == num[k+1]) k--;
                    }
                }
            }
            return ret;
        }

};
int main(int argc, char* argv[]){
}
